components/core/src/crypto/keys/mod.rs

```diff
@@ -287,6 +291,20 @@ where
             }
         };
 
+        let filename = match dir_entry.file_name().into_string() {
+            Ok(f) => f,
+            Err(e) => {
+                // filename is still an OsString, so print it as debug output
+                debug!("Invalid filename {:?}", e);
+                return Err(Error::CryptoError(format!("Invalid filename in key path")));
+            }
+        };
+
+        debug!("checking file: {}", &filename);
+        if !check_filename(keyname, filename, &mut candidates, pair_type) {
+            continue;
+        }
+
         let file = File::open(dir_entry.path())?;
         let mut reader = BufReader::new(file);
         let mut buf = String::new();
```

Comment:
I can't comment on the line I want since it's not in the diff... Line 311 in this patch:

```rust
reader.read_line(&mut buf)?;
```

This is where the error is coming from, because `read_line` [can fail when the content is not UTF-8](https://doc.rust-lang.org/1.21.0/std/io/trait.BufRead.html#method.read_line). By explicitly handling this error (instead of passing it out using the `?` operator), we'll fix this immediate issue due to MacOS `.DS_Store` files, but will also handle any other file that happens to have non-UTF-8 data.

Moving the filename checking to before the read call, as done in this PR, is a welcome change (it allows you to catch things earlier on), but handling this error is the core of the issue, and will also make Habitat more robust.

Reply:
How could I write this exactly? With a match that would lead to a continue if errored?

Reply:
@rsertelon Yes, exactly! You're in a loop populating the vector of candidates right now, but this error is just for one of them. The code currently bails out of everything if one file is bad, which is why this is a bug in the first place 😄 . You just want to say "yup, that's a bad file", and go on to process the rest of your files, and ultimately return a list of candidates.

Once we get into that candidate processing loop, we should be handling all possible errors.

Reply:
Fixed the error handling, but not sure about the implementation (I don't like returning the contents of `Ok()`)